FEC-10339 exo player v2.12.0 upgrade

giladna opened the pull request:
### Description of the Changes

Please add a detailed description of the change, whether it's an enhancement or a bugfix.
If the PR is related to an open issue please link to it.

### CheckLists

- [ ] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment 
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

giladna committed:
remove ExoMediaCrypto

giladna committed:
using exo 2.12.0

giladna committed:
Merge branch 'dev' into FEC-10339

* dev:
  FEC-10282 Preference for internal and external subtitles (#644)
  - Added nullability check while accessing preferredTextTrackConfig / preferredAudioTrackConfig (#651)
  - Clearing audioTracksCodecsMap on change media (#650)
  - Expose API for HEVC (hardware/software) capabilities (#648)
  FEC-10452 - Expose missing API setLicenseRequestAdapter  (#649)
  FEC-10340 Multi-codec support for Audio/Video (#647)

giladna committed:
exo player 2.12 upgrade cont

giladna committed:
finalize exo 2.12 upgrade phase 1

next phase add new api useage

giladna committed:
add parameter getDrmLicenseUrl

giladna committed:
remove unused drmSessionManager

giladna committed:
-public class PKDeviceSupportInfo
+public class PKDeviceCapabilitiesInfo

player.setAudioAttributes(AudioAttributes.DEFAULT, /* handleAudioFocus= */ true);

giladna committed:
add LocalExoMediaSource

giladna committed:
expose cache to playkit
drm support in cache

giladna committed:
Merge branch 'dev' into FEC-10339
add multicast
use mediasource with item constructor

# Conflicts:
#	playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java

giladna committed:
remove empty line

giladna committed:
fix typo

giladna committed:
Merge branch 'FEC-10339' of https://github.com/kaltura/playkit-android into FEC-10339

giladna committed:
fix wrong refarance

giladna committed:
add javadoc

giladna committed:
change to one line

giladna committed:
remove empty line

giladna committed:
cosmetics

giladna committed:
add
onIsLoadingChanged

giladna committed:
set cache only if not null

giladna committed:
add missing license headers

giladna committed:
protect npe

giladna committed:
use allowClearLead

giladna committed:
fix dtg playback
reuse mediaSourceFactory

giladna committed:
dtg offline playback

giladna committed:
fix external subtitles

giladna committed:
2.12.100 built from 2.12.0 fixes MediaItem.Subtitle do not contain label - https://github.com/giladna/ExoPlayer/tree/v2.12.0.1
fix missing label issue

giladna committed:
ident

noamtamim commented on playkit/build.gradle:
```diff
@@ -35,17 +35,18 @@ tasks.withType(Javadoc) {
 dependencies {
 
-    def exoPlayerVersion = '2.11.7'
+    def exoPlayerVersion = '2.12.100'
```
Why 100 and not 0?

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/LocalAssetsManagerExo.java:
```diff
@@ -59,6 +61,15 @@ public void unregisterAsset(String assetId, @Nullable byte[] drmInitData) {
         }
     }
 
+    /**
+     * @param assetId        - the id of the asset.
+     * @param exoMediaItem - the actual url of the video that should be played.
+     * @return - the {@link PKMediaSource} that should be passed to the player.
+     */
+    public PKMediaSource getLocalMediaSource(@NonNull final String assetId, @NonNull final MediaItem exoMediaItem) {
+        return new LocalExoMediaItem(helper.localDataStore, exoMediaItem, assetId, helper.getLocalAssetScheme(assetId));
+    }
+
```
The changes in this file are related to exo upgrade?

giladna replied:
yes related both for Kaltura player and dtg

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/Player.java:
```diff
@@ -15,6 +15,7 @@
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 
+import com.kaltura.android.exoplayer2.upstream.cache.Cache;
```
No Exo classes in our generic public API.

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/drm/DeferredDrmSessionManager.java:
```diff
@@ -69,15 +71,19 @@ public DeferredDrmSessionManager(Handler mainHandler, DrmCallback drmCallback, D
+//    public void setDrmSessionManager(DrmSessionManager drmSessionManager) {
+//        this.drmSessionManager = drmSessionManager;
+//    }
+
```
Remove?

giladna replied:
kept it there not in use right now

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/drm/DeferredDrmSessionManager.java:
```diff
@@ -87,41 +93,39 @@ public void setMediaSource(PKMediaSource mediaSource) {
+//            if (drmSessionManager instanceof DefaultDrmSessionManager) {
+//                ((DefaultDrmSessionManager) drmSessionManager).addListener(mainHandler, this);
+//            }
+//        }
```
Not needed anymore?

giladna replied:
removed

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/MediaSupport.java:
```diff
@@ -42,6 +42,7 @@
     private static final PKLog log = PKLog.get("MediaSupport");
 
     public static final UUID WIDEVINE_UUID = UUID.fromString("edef8ba9-79d6-4ace-a3c8-27dcd51d21ed");
+    public static final UUID PLAYREADY_UUID = UUID.fromString("9a04f079-9840-4286-ab92-e65be0885f95");
```
Why adding this?

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java:
```diff
@@ -280,7 +280,7 @@ private PKTracks buildTracks() {
         int defaultAudioTrackIndex = getDefaultTrackIndex(filteredAudioTracks, lastSelectedTrackIds[TRACK_TYPE_AUDIO]);
         int defaultTextTrackIndex = getDefaultTrackIndex(textTracks, lastSelectedTrackIds[TRACK_TYPE_TEXT]);
-
+        Collections.sort(videoTracks);
```
What are you sorting by, and why?

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java:
```diff
@@ -143,15 +149,15 @@
 
     private Timeline.Period period;
 
+    private Cache downloadCache;
```
Probably `@Nullable`

giladna replied:
removed for now

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java:
```diff
@@ -379,104 +382,127 @@ private void addExternalTextTrackErrorListener() {
 
         if (format == null) {
-            return null;
+            return null; // throw new IllegalArgumentException("Unknown media format: " + format + " for url: " + requestParams.url);
```
To throw or not to throw?

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java:
```diff
@@ -379,104 +382,127 @@ private void addExternalTextTrackErrorListener() {
+        MediaSource mediaSource;
         switch (format) {
             case dash:
+                if (sourceConfig.mediaSource.hasDrmParams()) {
```
This `case` is too large - extract method.

giladna committed:
Merge branch 'dev' into FEC-10339

giladna committed:
fix player restore position

giladna committed:
fix player restore position

giladna committed:
fix player restore position

giladna committed:
cleanup

giladna committed:
player.setAudioAttributes(AudioAttributes.DEFAULT, /* handleAudioFocus= */ playerSettings.isHandleAudioFocus());
        player.setHandleAudioBecomingNoisy(playerSettings.isHandleAudioBecomingNoisyEnabled());

giladna committed:
spaces

giladna committed:
handleAudioFocus = false as default

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerView.java:
```diff
@@ -83,20 +85,24 @@
+
+            @Override
+            public void onPlayWhenReadyChanged(boolean playWhenReady, int reason) {
+                isPlayWhenReady = playWhenReady;
```
check if we can ask player.isPlayWhenReady()...

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/ExternalTextTrackLoadErrorPolicy.java:
```diff
@@ -26,29 +26,30 @@ public void setOnTextTrackErrorListener(ExternalTextTrackLoadErrorPolicy.OnTextT
+                            log.e("Error-Event sent, type = " + PKPlayerErrorType.SOURCE_ERROR);
+                            textTrackLoadErrorListener.onTextTrackLoadError(currentError);
+                        }
+                        return Consts.TIME_UNSET;
```
Consider early exist instead of nested ifs.

giladna committed:
remove cache

giladna committed:
use player.getPlayWhenReady() instead of saving state

giladna committed:
add npe protection

giladna committed:
code review changes

giladna committed:
fix the npe

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/profiler/ExoPlayerProfilingListener.java:
```diff
@@ -127,7 +127,13 @@ public void onPlayerStateChanged(EventTime eventTime, boolean playWhenReady, int
+    @Override
+    public void onPlayWhenReadyChanged(EventTime eventTime, boolean playWhenReady, int reason) {
+        log("onPlayWhenReadyChanged", field("shouldPlay", playWhenReady));
+
```
Add member `shouldPlay` and log it

giladna committed:
use shouldPlay Flag

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java:
```diff
@@ -379,104 +364,96 @@ private void addExternalTextTrackErrorListener() {
+                            tracks.add(C.TRACK_TYPE_VIDEO);
+                            tracks.add(C.TRACK_TYPE_AUDIO);
+                            builder.setDrmSessionForClearTypes(tracks);
+                        }
```
Remove please

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java:
```diff
@@ -379,104 +364,96 @@ private void addExternalTextTrackErrorListener() {
-    }
-
-    private HttpDataSource.Factory getHttpDataSourceFactory(Map<String, String> headers) {
+    private HttpDataSource.Factory getHttpDataSourceFactory(Map<String,String> headers) {
```
```suggestion
    private HttpDataSource.Factory getHttpDataSourceFactory(Map<String, String> headers) {
```

giladna replied:
done

noamtamim commented on playkit/src/main/java/com/kaltura/playkit/player/ExoPlayerWrapper.java:
```diff
@@ -622,6 +596,12 @@ public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
         }
     }
 
+    @Override
+    public void onPlayWhenReadyChanged(boolean playWhenReady, int reason) {
+        log.d("onPlayWhenReadyChanged. playWhenReady = " + playWhenReady);
+
+    }
+
```
```suggestion
```

giladna replied:
done

giladna committed:
code review changes

giladna committed:
code review changes

giladna committed:
remove line

giladna committed:
remove MulticastSettings
